File: DP/LongestCommonSubsequence.cpp
#include <string>
#include <vector>
using namespace std;

/*
This is a very important question: https://leetcode.com/problems/longest-common-subsequence/description/
- The fundamential idea is a 2D table. Every pair of a substring, we are looking to find the longest common subsequence between those
- this problem lays a foundation for many other problems, and is core problem in DP
*/
// make 2D DP table

class Solution {
public:
  int longestCommonSubsequence(string text1, string text2) {
    int n = text1.size();
    int m = text2.size();
    vector<vector<int>> dp(n + 1, vector<int>(m + 1, 0));

    for (int i = n - 1; i >= 0; i--) {
      for (int j = m - 1; j >= 0; j--) {
        if (text1[i] == text2[j]) {
          dp[i][j] = 1 + dp[i + 1][j + 1];
        } else {
          dp[i][j] = max(dp[i][j + 1], dp[i + 1][j]);
        }
      }
    }
    return dp[0][0];
  }
};

File: rust_sols/src/dynamic_programming/longest_common_subsequence.rs
use std::cmp::max;

pub struct Solution {}

impl Solution {
    pub fn longest_common_subsequence(text1: String, text2: String) -> i32 {
        let text1: &[u8] = text1.as_bytes();
        let text2: &[u8] = text2.as_bytes();
        let n = text1.len();
        let m = text2.len();
        let mut dp: Vec<Vec<i32>> = vec![vec![0; m + 1]; n + 1];

        for i in (0..n).rev() {
            for j in (0..m).rev() {
                if text1[i] == text2[j] {
                    dp[i][j] = 1 + dp[i + 1][j + 1];
                } else {
                    dp[i][j] = max(dp[i + 1][j], dp[i][j + 1]);
                }
            }
        }
        dp[0][0]
    }
}
